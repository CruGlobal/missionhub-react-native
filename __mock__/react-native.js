import { NativeModules } from 'react-native';

// From https://github.com/facebook/react-native/issues/18279#issuecomment-374177940
// Will be made unnecessary once https://github.com/facebook/react-native/pull/18718 is released. It doesn't seem to be in 0.55.3
NativeModules.BlobModule = {
  ...NativeModules.BlobModule,
  addNetworkingHandler: jest.fn(),
};
