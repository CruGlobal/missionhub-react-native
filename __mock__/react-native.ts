jest.mock(
  'react-native/Libraries/Utilities/NativePlatformConstantsIOS',
  () => ({
    // @ts-ignore
    ...jest.requireActual(
      'react-native/Libraries/Utilities/NativePlatformConstantsIOS',
    ),
    getConstants: () => ({
      forceTouchAvailable: false,
      interfaceIdiom: 'en',
      isTesting: false,
      osVersion: 'ios',
      reactNativeVersion: { major: 60, minor: 1, patch: 0 },
      systemName: 'ios',
    }),
  }),
);
