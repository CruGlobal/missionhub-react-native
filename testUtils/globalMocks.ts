import faker from 'faker/locale/en';
import { IMocks } from 'graphql-tools';
import moment from 'moment';

import {
  CommunityCelebrationCelebrateableEnum,
  PermissionEnum,
  ReminderTypeEnum,
  StepTypeEnum,
} from '../__generated__/globalTypes';

let currentId = 1;
const nextId = () => currentId++;
export const resetGlobalMockSeeds = () => {
  currentId = 1;
  faker.seed(1);
};
resetGlobalMockSeeds();

export const globalMocks: IMocks = {
  String: () => faker.lorem.words(),
  Int: () => faker.random.number(),
  Float: () => faker.random.number({ precision: 0.01 }),
  Boolean: () => faker.random.boolean(),
  ID: () => nextId(),
  ISO8601DateTime: () => faker.date.past(10, '2020-01-14').toISOString(),
  ISO8601Date: () =>
    moment(faker.date.past(10, '2020-01-14')).format('YYYY-MM-DD'),

  BasePageInfo: () => ({
    endCursor: null,
    hasNextPage: false,
    hasPreviousPage: false,
    startCursor: null,
  }),
  ReminderTypeEnum: () => ReminderTypeEnum.once,
  StepTypeEnum: () => faker.random.arrayElement(Object.values(StepTypeEnum)),

  Step: () => ({
    title: faker.lorem.sentence(),
  }),
  Person: () => {
    const firstName = faker.name.firstName();
    const lastName = faker.name.lastName();
    return {
      fullName: `${firstName} ${lastName}`,
      firstName,
      lastName,
    };
  },
  Community: () => ({
    name: faker.company.catchPhrase(),
  }),
  CommunityCelebrationItem: () => {
    const firstName = faker.name.firstName();
    const lastName = faker.name.lastName();
    return {
      celebrateableType: faker.random.arrayElement(
        Object.values(CommunityCelebrationCelebrateableEnum),
      ),
      changedAttributeValue: moment(
        faker.date.past(10, '2020-01-14'),
      ).toISOString(),
      subjectPersonName: `${firstName} ${lastName}`,
    };
  },
  CommunityPermission: () => {
    return {
      permission: faker.random.arrayElement(Object.values(PermissionEnum)),
    };
  },
  FeedItem: () => {
    const firstName = faker.name.firstName();
    const lastName = faker.name.lastName();
    return {
      subjectPersonName: `${firstName} ${lastName}`,
    };
  },
};
