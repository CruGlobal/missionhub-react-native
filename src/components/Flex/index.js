import React, { Component } from 'react';
import { View } from 'react-native';
import PropTypes from 'prop-types';
import * as Animatable from 'react-native-animatable';

export default class Flex extends Component {
  setNativeProps(nativeProps) {
    this._view.setNativeProps(nativeProps);
  }
  render() {
    const { value, direction, align, justify, self: flexSelf, grow, wrap, children, style = {}, animation, ...rest } = this.props;
    let styleObj = {};
    if (value) styleObj.flex = value;
    if (direction) styleObj.flexDirection = direction;
    if (wrap) styleObj.flexWrap = wrap;
    if (grow) styleObj.flexGrow = grow;

    if (align) {
      if (align === 'center') styleObj.alignItems = 'center';
      else if (align === 'start') styleObj.alignItems = 'flex-start';
      else if (align === 'end') styleObj.alignItems = 'flex-end';
      else if (align === 'stretch') styleObj.alignItems = 'stretch';
    }
    if (flexSelf) {
      if (flexSelf === 'center') styleObj.alignSelf = 'center';
      else if (flexSelf === 'start') styleObj.alignSelf = 'flex-start';
      else if (flexSelf === 'end') styleObj.alignSelf = 'flex-end';
      else if (flexSelf === 'stretch') styleObj.alignSelf = 'stretch';
    }
    if (justify) {
      if (justify === 'center') styleObj.justifyContent = 'center';
      else if (justify === 'start') styleObj.justifyContent = 'flex-start';
      else if (justify === 'end') styleObj.justifyContent = 'flex-end';
      else if (justify === 'around') styleObj.justifyContent = 'space-around';
      else if (justify === 'between') styleObj.justifyContent = 'space-between';
    }
    // If animation is passed in, use the animatable library, otherwise don't
    if (animation) {
      return (
        <Animatable.View
          ref={(c) => this._view = c}
          duration={400}
          animation={animation}
          {...rest}
          style={[style, styleObj]}
        >
          {children}
        </Animatable.View>
      );
    }
    return (
      <View
        ref={(c) => this._view = c}
        {...rest}
        style={[style, styleObj]}
      >
        {children}
      </View>
    );
  }
}

Flex.propTypes = {
  children: PropTypes.node,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.number, PropTypes.array]),
  value: PropTypes.number,
  grow: PropTypes.number,
  direction: PropTypes.oneOf(['row', 'column']),
  wrap: PropTypes.oneOf(['wrap', 'wrap-reverse', 'nowrap']),
  align: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  justify: PropTypes.oneOf(['start', 'center', 'end', 'around', 'between']),
  self: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
};
