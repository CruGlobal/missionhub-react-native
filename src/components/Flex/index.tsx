/* eslint complexity: 0, max-lines-per-function: 0 */
import React, { forwardRef } from 'react';
import { View, StyleProp, ViewStyle } from 'react-native';
import * as Animatable from 'react-native-animatable';
import { ReactNode } from 'react';
import { Ref } from 'react';

interface FlexProps {
  style?: StyleProp<ViewStyle>;
  value?: number;
  grow?: number;
  direction?: 'row' | 'column';
  wrap?: 'wrap' | 'wrap-reverse' | 'nowrap';
  align?: 'start' | 'center' | 'end' | 'stretch';
  justify?: 'start' | 'center' | 'end' | 'around' | 'between';
  self?: 'start' | 'center' | 'end' | 'stretch';
  animated?: boolean;
  children?: ReactNode;
  animation?: string;
}

const Flex = forwardRef(
  (
    {
      value,
      direction,
      align,
      justify,
      self: flexSelf,
      grow,
      wrap,
      children,
      style = {},
      animation,
      animated,
      ...rest
    }: FlexProps,
    ref: Ref<View>,
  ) => {
    const styleObj: StyleProp<ViewStyle> = {};
    if (value) {
      styleObj.flex = value;
    }
    if (direction) {
      styleObj.flexDirection = direction;
    }
    if (wrap) {
      styleObj.flexWrap = wrap;
    }
    if (grow) {
      styleObj.flexGrow = grow;
    }
    if (align) {
      if (align === 'center') {
        styleObj.alignItems = 'center';
      } else if (align === 'start') {
        styleObj.alignItems = 'flex-start';
      } else if (align === 'end') {
        styleObj.alignItems = 'flex-end';
      } else if (align === 'stretch') {
        styleObj.alignItems = 'stretch';
      }
    }
    if (flexSelf) {
      if (flexSelf === 'center') {
        styleObj.alignSelf = 'center';
      } else if (flexSelf === 'start') {
        styleObj.alignSelf = 'flex-start';
      } else if (flexSelf === 'end') {
        styleObj.alignSelf = 'flex-end';
      } else if (flexSelf === 'stretch') {
        styleObj.alignSelf = 'stretch';
      }
    }
    if (justify) {
      if (justify === 'center') {
        styleObj.justifyContent = 'center';
      } else if (justify === 'start') {
        styleObj.justifyContent = 'flex-start';
      } else if (justify === 'end') {
        styleObj.justifyContent = 'flex-end';
      } else if (justify === 'around') {
        styleObj.justifyContent = 'space-around';
      } else if (justify === 'between') {
        styleObj.justifyContent = 'space-between';
      }
    }
    // If animation is passed in, use the animatable library, otherwise don't
    if (animation || animated) {
      return (
        <Animatable.View
          ref={ref}
          duration={400}
          animation={animation}
          {...rest}
          style={[style, styleObj]}
        >
          {children}
        </Animatable.View>
      );
    }
    return (
      <View ref={ref} {...rest} style={[style, styleObj]}>
        {children}
      </View>
    );
  },
);

Flex.displayName = 'Flex';

export default Flex;
