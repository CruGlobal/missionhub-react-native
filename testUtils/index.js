import 'react-native';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Need to import for all translation screens to work properly
import '../src/i18n';

Enzyme.configure({ adapter: new Adapter() });

export const createMockStore = (state = {}) => {
  return {
    getState: jest.fn(() => (state)),
    dispatch: jest.fn((response) => Promise.resolve(response)),
    subscribe: jest.fn(),
  };
};

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (data) => {
  expect(renderer.create(data)).toMatchSnapshot();
};

export const renderShallow = (component, store) => {
  let renderedComponent = shallow(
    component,
    { context: { store: store } }
  );

  // If component has translation wrappers, dive deeper
  while (renderedComponent.is('Translate') || renderedComponent.is('I18n')) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (component, store) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};

export const mockFnWithParams = (obj, method, expectedReturn, ...expectedParams) => {
  return obj[method] = jest.fn(
    (...actualParams) => {
      expect(actualParams).toEqual(expectedParams);
      return expectedReturn;
    }
  );
};
