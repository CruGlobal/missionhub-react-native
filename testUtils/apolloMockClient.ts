import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema, IMocks } from 'graphql-tools';
import { DocumentNode } from 'graphql';
import { buildClientSchema, IntrospectionQuery } from 'graphql/utilities';
import gql from 'fraql';
import { createMockerFromIntrospection } from 'fraql/mock';

import introspectionQuery from '../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks: IMocks = {}) => {
  const schema = buildClientSchema(
    (introspectionQuery as unknown) as IntrospectionQuery,
  );

  addMockFunctionsToSchema({
    schema,
    mocks: { ...globalMocks, ...mocks },
  });

  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new SchemaLink({ schema }),
  });
};

export const mockFragment = <TData>(
  fragmentDocument: DocumentNode,
  options?: {
    mocks: IMocks;
  },
): TData => {
  if (!fragmentDocument.loc) {
    throw 'Unable to get original gql query to create FraQL fragment.';
  }

  return createMockerFromIntrospection(introspectionQuery, {
    mocks: globalMocks,
  }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
