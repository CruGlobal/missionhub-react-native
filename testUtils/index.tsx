import React, { ReactElement, ReactNode } from 'react';
import 'react-native';
import { Provider } from 'react-redux';
import { Provider as ProviderLegacy } from 'react-redux-legacy';
import thunk from 'redux-thunk';
import configureStore, { MockStore } from 'redux-mock-store';
// eslint-disable-next-line import/named
import { NavigationParams, NavigationProvider } from 'react-navigation';
import { ApolloProvider } from '@apollo/react-hooks';
import { ReactTestRendererJSON } from 'react-test-renderer';
import { render, RenderAPI } from 'react-native-testing-library';
import snapshotDiff from 'snapshot-diff';
import Enzyme, { shallow as enzymeShallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { IMocks } from 'graphql-tools';

import { createApolloMockClient } from './apolloMockClient';
import { createNavigationProp } from './navigationHelpers';

Enzyme.configure({ adapter: new Adapter() });

export const createThunkStore = configureStore([thunk]);

interface RenderWithContextParams {
  initialState?: {};
  store?: MockStore;
  navParams?: NavigationParams;
  mocks?: IMocks;
  noWrappers?: boolean;
}

// Inspiration from https://github.com/kentcdodds/react-testing-library/blob/52575005579307bcfbe7fbe4ef4636147c03c6fb/examples/__tests__/react-redux.js#L69-L80
export function renderWithContext(
  component: ReactElement,
  {
    initialState,
    store = createThunkStore(initialState),
    navParams,
    mocks: mocks = {},
    noWrappers = false,
  }: RenderWithContextParams = {},
) {
  const mockApolloClient = createApolloMockClient(mocks);

  const navigation = createNavigationProp(navParams);

  let renderResult: RenderAPI;
  if (noWrappers) {
    renderResult = render(React.cloneElement(component, { navigation }));
  } else {
    // Warning: don't call any functions in here that return new instances on every call. All the props need to stay the same otherwise rerender won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
      <NavigationProvider value={navigation}>
        <ProviderLegacy store={store}>
          <Provider store={store}>
            <ApolloProvider client={mockApolloClient}>
              {children}
            </ApolloProvider>
          </Provider>
        </ProviderLegacy>
      </NavigationProvider>
    );

    renderResult = render(React.cloneElement(component, { navigation }), {
      wrapper,
    });
  }

  let storedSnapshot: ReactTestRendererJSON | null;
  return {
    ...renderResult,
    store,
    snapshot: () => expect(renderResult.toJSON()).toMatchSnapshot(),
    recordSnapshot: () => (storedSnapshot = renderResult.toJSON()),
    diffSnapshot: () => {
      if (!storedSnapshot) {
        throw new Error(
          'You must call recordSnapshot to store an initial snapshot before calling diffSnapshot',
        );
      }
      expect(
        snapshotDiff(storedSnapshot, renderResult.toJSON()),
      ).toMatchSnapshot();
    },
  };
}

// TODO: Remove all legacy rendering functions below

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (component: ReactElement) => {
  const { toJSON } = render(component);
  expect(toJSON()).toMatchSnapshot();
};

export const renderShallow = (
  component: ReactElement,
  store = createThunkStore(),
) => {
  let renderedComponent = enzymeShallow(
    <ProviderLegacy store={store}>{component}</ProviderLegacy>,
  ).dive();

  // If component has translation wrappers, dive deeper
  while (
    // eslint-disable-next-line  @typescript-eslint/no-explicit-any
    ((renderedComponent.type() as any).displayName || '').startsWith(
      'withI18nextTranslation(',
    )
  ) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (
  component: ReactElement,
  store = createThunkStore(),
) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};
