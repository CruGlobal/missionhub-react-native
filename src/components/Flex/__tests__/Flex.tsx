import 'react-native';
import React from 'react';

import { testSnapshot } from '../../../../testUtils';

import Flex from '..';

it('renders correctly', () => {
  testSnapshot(<Flex />);
});

it('renders align center correctly', () => {
  testSnapshot(<Flex align="center" />);
});

it('renders justify center correctly', () => {
  testSnapshot(<Flex justify="center" />);
});

it('renders alignSelf center correctly', () => {
  testSnapshot(<Flex self="center" />);
});

it('renders value correctly', () => {
  testSnapshot(<Flex value={1} />);
});

it('renders direction correctly', () => {
  testSnapshot(<Flex direction="row" />);
});

it('renders animation correctly', () => {
  testSnapshot(<Flex animation="bounce" />);
});

it('renders grow correctly', () => {
  testSnapshot(<Flex grow={2} />);
});
