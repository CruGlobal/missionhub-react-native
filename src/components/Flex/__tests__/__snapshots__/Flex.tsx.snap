// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders align center correctly 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "alignItems": "center",
      },
    ]
  }
/>
`;

exports[`renders alignSelf center correctly 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "alignSelf": "center",
      },
    ]
  }
/>
`;

exports[`renders animation correctly 1`] = `
<View
  style={
    Object {
      "transform": Array [
        Object {
          "translateY": 0,
        },
      ],
    }
  }
/>
`;

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {},
      Object {},
    ]
  }
/>
`;

exports[`renders direction correctly 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "flexDirection": "row",
      },
    ]
  }
/>
`;

exports[`renders grow correctly 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "flexGrow": 2,
      },
    ]
  }
/>
`;

exports[`renders justify center correctly 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "justifyContent": "center",
      },
    ]
  }
/>
`;

exports[`renders value correctly 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "flex": 1,
      },
    ]
  }
/>
`;
